Promote the `wasm32-wasip2` target to Tier 2

alexcrichton opened the pull request:
This commit promotes the `wasm32-wasip2` Rust target to tier 2 as proposed in rust-lang/compiler-team#760. There are two major changes in this PR:

1. The `dist-various-2` container, which already produces the other WASI targets, now has an extra target added for `wasm32-wasip2`.
2. A new `wasm-component-ld` binary is added to all host toolchains when LLD is enabled. This is the linker used for the `wasm32-wasip2` target.

This new linker is added for all host toolchains to ensure that all host toolchains can produce the `wasm32-wasip2` target. This is similar to how `rust-lld` was originally included for all host toolchains to be able to produce WebAssembly output when the targets were first added. The new linker is developed [here][wasm-component-ld] and is pulled in via a crates.io-based dependency to the tree here.

[wasm-component-ld]: https://github.com/bytecodealliance/wasm-component-ld

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

alexcrichton committed:
Promote the `wasm32-wasip2` target to Tier 2

This commit promotes the `wasm32-wasip2` Rust target to tier 2 as
proposed in rust-lang/compiler-team#760. There are two major changes in
this PR:

1. The `dist-various-2` container, which already produces the other WASI
   targets, now has an extra target added for `wasm32-wasip2`.
2. A new `wasm-component-ld` binary is added to all host toolchains when
   LLD is enabled. This is the linker used for the `wasm32-wasip2` target.

This new linker is added for all host toolchains to ensure that all host
toolchains can produce the `wasm32-wasip2` target. This is similar to
how `rust-lld` was originally included for all host toolchains to be
able to produce WebAssembly output when the targets were first added.
The new linker is developed [here][wasm-component-ld] and is pulled in
via a crates.io-based dependency to the tree here.

[wasm-component-ld]: https://github.com/bytecodealliance/wasm-component-ld

alexcrichton commented:
I'll note that https://github.com/rust-lang/compiler-team/issues/760 is not complete so this shouldn't merge yet. I'm opening this up ahead of time to get feedback on the more technical side of things as opposed to the MCP-side of things.

One specific question I'd have is that this PR is adding more exceptions to the license check for crates that have an `Apache-2.0 WITH LLVM-exception` license. I understand it's easiest to stick to `MIT OR Apache-2.0` like most other Rust crates and such, and to that effect we're going to try to relicense the [wasm-tools repository](https://github.com/bytecodealliance/wasm-tools) to `Apache-2.0 WITH LLVM-exception OR MIT OR Apache-2.0` soon. In lieu of this though, as that will probably take a bit, two questions:

* Is it ok to add more exceptions here in the meantime?
* Is it necessary to go through the relicensing effort? There are a number of other crates such as `ar_archive_writer` and `rustc_apfloat` and Cranelift crates which have the same `Apache-2.0 WITH LLVM-exception`. I don't expect Wasmtime's own license (and thus Cranelift's) to change, it's just the `wasm-tools`-family of crates that may change. If this is an acceptable state, should we consider it ok to not relicense wasm-tools?

Also I realize that licensing decisions are probably best not made in PRs, so this can also be considered as mostly highlighting the issue and if you know a better forum/location to discuss this I'm happy to raise something there.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9668843568/job/26673979715) [(plain)](https://github.com/rust-lang/rust/commit/674267e0a6b443265692f3932efdc7a3f74a35d5/checks/26673979715/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling log v0.4.21
error[E0725]: the feature `min_specialization` is not in the list of allowed features
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/lib.rs:100:45
    |
100 | #![cfg_attr(feature = "specialize", feature(min_specialization))]

error[E0658]: specialization is unstable
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:43:5
   |
   |
43 | /     default fn get_hash<H: Hash + ?Sized, B: BuildHasher>(value: &H, build_hasher: &B) -> u64 {
44 | |         let mut hasher = build_hasher.build_hasher();
45 | |         value.hash(&mut hasher);
46 | |         hasher.finish()
   | |_____^
   |
   = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
   = help: add `#![feature(specialization)]` to the crate attributes to enable
   = help: add `#![feature(specialization)]` to the crate attributes to enable
   = note: this compiler was built on 2024-06-25; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/lib.rs:266:5
    |
266 | /     default fn hash_as_u64<T: Hash + ?Sized>(&self, value: &T) -> u64 {
267 | |         let mut hasher = self.build_hasher();
268 | |         value.hash(&mut hasher);
269 | |         hasher.finish()
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2024-06-25; consider upgrading it if it is out of date

   Compiling fastrand v2.1.0
error[E0658]: specialization is unstable
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/lib.rs:280:5
    |
280 | /     default fn hash_as_fixed_length<T: Hash + ?Sized>(&self, value: &T) -> u64 {
281 | |         let mut hasher = self.build_hasher();
282 | |         value.hash(&mut hasher);
283 | |         hasher.finish()
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2024-06-25; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/lib.rs:294:5
    |
294 | /     default fn hash_as_str<T: Hash + ?Sized>(&self, value: &T) -> u64 {
295 | |         let mut hasher = self.build_hasher();
296 | |         value.hash(&mut hasher);
297 | |         hasher.finish()
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2024-06-25; consider upgrading it if it is out of date

error[E0119]: conflicting implementations of trait `BuildHasherExt` for type `RandomState`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/lib.rs:263:1
    |
263 | impl<B: BuildHasher> BuildHasherExt for B {
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/random_state.rs:469:1
    |
469 | impl BuildHasherExt for RandomState {
469 | impl BuildHasherExt for RandomState {
    | ----------------------------------- first implementation here

error[E0119]: conflicting implementations of trait `CallHasher` for type `u128`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:71:1
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
...
71 |   impl CallHasher for u128 {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `u128`
error[E0119]: conflicting implementations of trait `CallHasher` for type `i128`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:79:1
   |
38 | / impl<T> CallHasher for T
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
79 |   impl CallHasher for i128 {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `i128`

error[E0119]: conflicting implementations of trait `CallHasher` for type `usize`
error[E0119]: conflicting implementations of trait `CallHasher` for type `usize`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:87:1
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
87 |   impl CallHasher for usize {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `usize`

error[E0119]: conflicting implementations of trait `CallHasher` for type `isize`
error[E0119]: conflicting implementations of trait `CallHasher` for type `isize`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:95:1
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
95 |   impl CallHasher for isize {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `isize`

error[E0119]: conflicting implementations of trait `CallHasher` for type `[u8]`
error[E0119]: conflicting implementations of trait `CallHasher` for type `[u8]`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:103:1
    |
38  | / impl<T> CallHasher for T
39  | | where
40  | |     T: Hash + ?Sized,
    | |_____________________- first implementation here
...
103 |   impl CallHasher for [u8] {

error[E0119]: conflicting implementations of trait `CallHasher` for type `Vec<u8>`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:111:1
    |
    |
38  | / impl<T> CallHasher for T
39  | | where
40  | |     T: Hash + ?Sized,
    | |_____________________- first implementation here
111 |   impl CallHasher for Vec<u8> {
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Vec<u8>`

error[E0119]: conflicting implementations of trait `CallHasher` for type `str`
error[E0119]: conflicting implementations of trait `CallHasher` for type `str`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:119:1
    |
38  | / impl<T> CallHasher for T
39  | | where
40  | |     T: Hash + ?Sized,
    | |_____________________- first implementation here
119 |   impl CallHasher for str {
    |   ^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `str`

error[E0119]: conflicting implementations of trait `CallHasher` for type `String`
error[E0119]: conflicting implementations of trait `CallHasher` for type `String`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:127:1
    |
38  | / impl<T> CallHasher for T
39  | | where
40  | |     T: Hash + ?Sized,
    | |_____________________- first implementation here
127 |   impl CallHasher for String {
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `String`

error[E0119]: conflicting implementations of trait `CallHasher` for type `u8`
error[E0119]: conflicting implementations of trait `CallHasher` for type `u8`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `u8`
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
   | |_- in this expansion of `call_hasher_impl!`
61 |   call_hasher_impl!(u8);

error[E0119]: conflicting implementations of trait `CallHasher` for type `u16`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `u16`
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
   | |_- in this expansion of `call_hasher_impl!`
61 |   call_hasher_impl!(u8);
62 |   call_hasher_impl!(u16);

error[E0119]: conflicting implementations of trait `CallHasher` for type `u32`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `u32`
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
...
...
63 |   call_hasher_impl!(u32);

error[E0119]: conflicting implementations of trait `CallHasher` for type `u64`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `u64`
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
...
...
64 |   call_hasher_impl!(u64);

error[E0119]: conflicting implementations of trait `CallHasher` for type `i8`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
...  |
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
   | |_- in this expansion of `call_hasher_impl!`
...
65 |   call_hasher_impl!(i8);

error[E0119]: conflicting implementations of trait `CallHasher` for type `i16`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
...  |
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
   | |_- in this expansion of `call_hasher_impl!`
...
66 |   call_hasher_impl!(i16);

error[E0119]: conflicting implementations of trait `CallHasher` for type `i32`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
...  |
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
   | |_- in this expansion of `call_hasher_impl!`
...
67 |   call_hasher_impl!(i32);

error[E0119]: conflicting implementations of trait `CallHasher` for type `i64`
  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.11/src/specialize.rs:53:9
   |
   |
38 | / impl<T> CallHasher for T
39 | | where
40 | |     T: Hash + ?Sized,
   | |_____________________- first implementation here
50 | / macro_rules! call_hasher_impl {
50 | / macro_rules! call_hasher_impl {
51 | |     ($typ:ty) => {
52 | |         #[cfg(feature = "specialize")]
53 | |         impl CallHasher for $typ {
...  |
59 | |     };
60 | | }
   | |_- in this expansion of `call_hasher_impl!`
   | |_- in this expansion of `call_hasher_impl!`
...
68 |   call_hasher_impl!(i64);

   Compiling clap_builder v4.5.7
Some errors have detailed explanations: E0119, E0658, E0725.
For more information about an error, try `rustc --explain E0119`.
For more information about an error, try `rustc --explain E0119`.
error: could not compile `ahash` (lib) due to 22 previous errors
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:10:37
+ set -e
+ cat /tmp/toolstate/toolstates.json
{"wasm-component-ld":"build-fail","lld-wrapper":"test-fail"}
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
##[endgroup]
ERROR: Tool `book` was not recorded in tool state.
ERROR: Tool `nomicon` was not recorded in tool state.
ERROR: Tool `reference` was not recorded in tool state.
ERROR: Tool `rust-by-example` was not recorded in tool state.
ERROR: Tool `edition-guide` was not recorded in tool state.
ERROR: Tool `embedded-book` was not recorded in tool state.
  local time: Tue Jun 25 20:17:26 UTC 2024
  network time: Tue, 25 Jun 2024 20:17:26 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Mark-Simulacrum commented:
> One specific question I'd have is that this PR is adding more exceptions to the license check for crates that have an Apache-2.0 WITH LLVM-exception license

My sense is that we're fine with extra dependencies with that license, especially in the same pool (i.e., wasm-related crates). cc @wesleywiser @davidtwco -- does that align with your expectations? Can you help drive the licensing story with the Foundation's help if needed?

I'm not sure whether re-licensing is worth the trouble or not as such. Though long-term it does likely hurt our ability to move code across the boundary of these crates, so if there's expectations that might be warranted it might be worth investing while there's less people involved (and so it's easier).

---

> This is similar to how rust-lld was originally included for all host toolchains to be able to produce WebAssembly output when the targets were first added. The new linker is developed [here](https://github.com/bytecodealliance/wasm-component-ld) and is pulled in via a crates.io-based dependency to the tree here.

Can you say how big the extra binary is? I don't remember the rationale back then, but it feels plausible that the linker should be part of the wasm target component... maybe I'm missing something obvious? (lld is distinct in that lots of targets use it -- at least now -- whereas this one seems like that may not be true for).

bors commented:
:umbrella: The latest upstream changes (presumably #127133) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

alexcrichton commented:
Ok sounds good on the licensing front. I opened https://github.com/bytecodealliance/wasm-tools/issues/1637 and https://github.com/bytecodealliance/wit-bindgen/issues/987 earlier anyway to start the process and see generally if it can be done, so that's progressing in parallel too.

As for binary sizes the linker locally is ~5M (4.6 on macos and 5.2 on linux). Making this part of the wasm component would require having a binary for each supported host architeture in the component as well which I think would be pretty difficult build-wise and also would make the wasm component significantly larger. IIRC we wrestled with this at the time when adding rust-lld as well and the decision was to ship it with all host components as that was the simplest solution at the time. Now that was also guided by the fact that we wanted to eventually use LLD for more targets one day (as has since happened) and this won't ever be used for any other target. I can look into size-optimizing the binary if 5M is too large.

alexcrichton committed:
Update how wasm-component-ld is built

Reuse preexisting macro and switch it to a "bootstrap tool" to try to
resolve build issues.

wesleywiser commented:
> My sense is that we're fine with extra dependencies with that license, especially in the same pool (i.e., wasm-related crates). cc @wesleywiser @davidtwco -- does that align with your expectations? Can you help drive the licensing story with the Foundation's help if needed?

Yes, I think that's ok. Obviously, it would be more ideal if we had the standard licensing applied so relicensing would still be preferable but not a blocker. Thanks @alexcrichton for starting that conversation in the bytecodealliance projects!

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9862845157/job/27234324897) [(plain)](https://github.com/rust-lang/rust/commit/1afdd45881d39f6af434fba77ae07e68e1b62bce/checks/27234324897/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
thread 'main' panicked at src/core/builder.rs:1500:9:
assertion failed: !use_snapshot || stage == 0 || self.local_rebuild
Build completed unsuccessfully in 0:00:00
+ set -e
+ cat /tmp/toolstate/toolstates.json
cat: /tmp/toolstate/toolstates.json: No such file or directory
```

</details>

alexcrichton committed:
Unconditionally use stage0 build compiler

alexcrichton commented:
Ok I've redone a bit how the build is done to fix the build error from before, notably unconditionally using the stage0 compiler for the wasm-component-ld linker to ensure that it avoids the nightly issue with the ahash dependency. The FCP has finished here and otherwise the relicensing efforts are underway and should hopefully be complete in a month or so with new published versions at which point I'll update this to remove the exceptions.

In the meantime this should otherwise be ready for review and I'm happy to dig further into the binary size if that's a concern. The easiest win is to probably compile with `-Cpanic=abort` which I can try to figure out how to plumb through if desired.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 4cd6eee894ec87de26b8dc212363c23f1db70c62 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 4cd6eee894ec87de26b8dc212363c23f1db70c62 -->
<!-- homu: {"type":"Approved","sha":"4cd6eee894ec87de26b8dc212363c23f1db70c62","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Mark-Simulacrum commented:
Let's land this. I think we can update the licensing and consider further improvements in binary size in the future, no need to block on that.

workingjubilee commented:
There was a failure in https://github.com/rust-lang/rust/pull/127696#issuecomment-2227132583 but it's not clear which PR failed it. Marking all target-specific and CI-related PRs from that rollup as iffy for now.

@bors rollup=iffy

bors commented:
:hourglass: Testing commit 4cd6eee894ec87de26b8dc212363c23f1db70c62 with merge f24522ae7b6427c2ca94da8e493e0126025acbda...
<!-- homu: {"type":"BuildStarted","head_sha":"4cd6eee894ec87de26b8dc212363c23f1db70c62","merge_sha":"f24522ae7b6427c2ca94da8e493e0126025acbda"} -->

rust-log-analyzer commented:
The job **`dist-arm-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9924797882/job/27416516093) [(plain)](https://github.com/rust-lang-ci/rust/commit/f24522ae7b6427c2ca94da8e493e0126025acbda/checks/27416516093/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] syn test:false 3.561
   Compiling semver v1.0.23
error[E0463]: can't find crate for `core`
  |
  = note: the `arm-unknown-linux-gnueabi` target may not be installed
  = help: consider downloading the target with `rustup target add arm-unknown-linux-gnueabi`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`
   Compiling equivalent v1.0.1
For more information about this error, try `rustc --explain E0463`.
[RUSTC-TIMING] cfg_if test:false 0.022
error: could not compile `cfg-if` (lib) due to 1 previous error
error: could not compile `cfg-if` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] once_cell test:false 0.026
error: could not compile `once_cell` (lib) due to 1 previous error
error[E0463]: can't find crate for `std`
  |
  = note: the `arm-unknown-linux-gnueabi` target may not be installed
  = help: consider downloading the target with `rustup target add arm-unknown-linux-gnueabi`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`
[RUSTC-TIMING] bitflags test:false 0.031
error: could not compile `bitflags` (lib) due to 1 previous error
[RUSTC-TIMING] zerocopy test:false 0.036
error: could not compile `zerocopy` (lib) due to 1 previous error
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9924797882/job/27416849739)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9924797882/job/27416849739","builder_name":"checks-actions"} -->

alexcrichton commented:
Definitely a bug with this PR. I changed this to always use the stage0 compiler but that doesn't actually work because off cross-compiling the compiler to a target that isn't the same as the stage0 compiler. That means I'll have to use a non-stage0 compiler which brings back the original issue with `ahash`. Looks like the `ahash` crate maintainer isn't around at this time though so I'll try to work through an alternative solution.

alexcrichton committed:
Don't use stage0 compiler for wasm-component-ld

Switch it to using a just-built standard library which enables it to be
cross compiled. Additionally allow it access to `min_specialization`
which `ahash`, a dependency, wants.

alexcrichton commented:
Ok I've updated how the tool is built, notably classifying it as `ToolStd` instead of `ToolBuild` which enables cross-compiling the binary for distribution. I'll also note that I've had to allow it access to `min_specialization` as a nightly feature. I've opened https://github.com/tkaitchuck/aHash/issues/238 to remove the need for thise in addition to sending a PR, but I haven't gotten a response yet.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit c370bf44d84a2e83617712de81ac602f59acc63a has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum c370bf44d84a2e83617712de81ac602f59acc63a -->
<!-- homu: {"type":"Approved","sha":"c370bf44d84a2e83617712de81ac602f59acc63a","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
This has regressed the bootstrap build time by ~70s (https://github.com/rust-lang/rust/pull/127831#issuecomment-2232518920), and it is being built on every build of the compiler that has "LLD in sysroot" enabled. The same happens with `lld-wrapper`, but that is a 100 line program without dependencies, while the `wasm-component-ld` tool has tens of dependencies and takes a while to build.

I don't care *that* much about the bootstrap time on perf.RLO, but this could also be slowing down local rustc rebuilds. Maybe we should create a separate `config.toml` flag for this to make it optional?

alexcrichton commented:
Sure thing, [happy to do that](https://github.com/rust-lang/rust/pull/127866)

RalfJung commented on src/tools/wasm-component-ld/README.md:
```diff
@@ -0,0 +1,62 @@
+# `wasm-component-ld`
+
+This wrapper is a wrapper around the [`wasm-component-ld`] crates.io crate. That
+crate. That crate is itself a thin wrapper around two pieces:
```
There's an incomplete sentence here (and still seems to be there on master).

alexcrichton replied:
Oh oops, I'll work on fixing this.
